lib/vrl/stdlib/src/parse_query_string.rs

```diff
@@ -0,0 +1,131 @@
+use bytes::Buf;
+use std::collections::btree_map::Entry;
+use std::collections::BTreeMap;
+use url::form_urlencoded;
+use vrl::prelude::*;
+
+#[derive(Clone, Copy, Debug)]
+pub struct ParseQueryString;
+
+impl Function for ParseQueryString {
+    fn identifier(&self) -> &'static str {
+        "parse_query_string"
+    }
+
+    fn examples(&self) -> &'static [Example] {
+        &[Example {
+            title: "parse query string",
+            source: r#"parse_query_string("foo=1&bar=2")"#,
+            result: Ok(r#"
+                {
+                    "foo": "1",
+                    "bar": "2"
+                }
+            "#),
+        }]
+    }
+
+    fn compile(&self, mut arguments: ArgumentList) -> Compiled {
+        let value = arguments.required("value");
+        Ok(Box::new(ParseQueryStringFn { value }))
+    }
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::BYTES,
+            required: true,
+        }]
+    }
+}
+
+#[derive(Debug, Clone)]
+struct ParseQueryStringFn {
+    value: Box<dyn Expression>,
+}
+
+impl Expression for ParseQueryStringFn {
+    fn resolve(&self, ctx: &mut Context) -> Resolved {
+        let query_string = self.value.resolve(ctx)?.try_bytes()?;
+        let mut result: BTreeMap<String, Value> = BTreeMap::new();
+        let parsed = form_urlencoded::parse(query_string.bytes());
+        for (k, v) in parsed {
+            let value = v.as_ref().into();
+            let entry = result.entry(k.as_ref().to_owned());
+            match entry {
+                Entry::Occupied(mut e) => {
+                    if e.get().is_array() {
+                        e.get_mut().as_array_mut().unwrap().push(value);
+                    } else {
+                        let prev_value = e.get().to_owned();
+                        result.insert(k.as_ref().into(), vec![prev_value, value].into());
+                    }
+                }
+                Entry::Vacant(e) => {
+                    e.insert(value);
+                }
+            }
```

Comment:
```suggestion
            result
                .entry(k.into_owned())
                .and_modify(|v| match v {
                    Value::Array(v) => v.push(value),
                    v => value!([v, value]),
                })
                .or_insert(value);
```

_Something_ like this should work, and seems a bit nicer to read?

Reply:
Much nicer, I remember I was struggling to make it work first and gave up thinking it is impossible, now I hope I got it right - @JeanMertz could you take a look again please to make sure it follows rust conventions/is optimal?